main/lsp/updates.cc

```diff
@@ -126,56 +147,84 @@ void tryApplyDefLocSaver(const core::GlobalState &gs, vector<ast::ParsedFile> &i
+                    auto &ast = pair.second;
+                    if (ast.tree) {
+                        indexedCopies.emplace_back(ast::ParsedFile{ast.tree->deepCopy(), ast.file});
+                        updatedFiles.insert(ast.file.id());
+                    }
+                    updates.updatedFinalGSFileIndexes.push_back(move(ast));
+                }
             }
-            updates.updatedFinalGSFileIndexes.push_back(move(ast));
-        }
-    }
 
-    // Copy the indexes of unchanged files.
-    for (const auto &tree : indexed) {
-        // Note: indexed entries for payload files don't have any contents.
-        if (tree.tree && !updatedFiles.contains(tree.file.id())) {
-            indexedCopies.emplace_back(ast::ParsedFile{tree.tree->deepCopy(), tree.file});
-        }
-    }
+            // Copy the indexes of unchanged files.
+            for (const auto &tree : indexed) {
+                // Note: indexed entries for payload files don't have any contents.
+                if (tree.tree && !updatedFiles.contains(tree.file.id())) {
+                    indexedCopies.emplace_back(ast::ParsedFile{tree.tree->deepCopy(), tree.file});
+                }
+            }
+            if (finalGS->isTypecheckingCanceled()) {
+                return false;
+            }
+
+            ENFORCE(finalGS->lspQuery.isEmpty());
+            auto maybeResolved =
+                pipeline::resolve(finalGS, move(indexedCopies), config.opts, workers, config.skipConfigatron);
+            if (!maybeResolved.hasResult()) {
+                return false;
+            }
+
+            auto &resolved = maybeResolved.result();
+            for (auto &tree : resolved) {
+                ENFORCE(tree.file.exists());
+                affectedFiles.push_back(tree.file);
+            }
+            auto maybeTypechecked = pipeline::typecheck(finalGS, move(resolved), config.opts, workers);
+            if (!maybeTypechecked.hasResult()) {
+                return false;
+            }
+            return true;
+        });
 
-    ENFORCE(finalGS->lspQuery.isEmpty());
-    auto resolved = pipeline::resolve(finalGS, move(indexedCopies), config.opts, workers, config.skipConfigatron);
-    vector<core::FileRef> affectedFiles;
-    for (auto &tree : resolved) {
-        ENFORCE(tree.file.exists());
-        affectedFiles.push_back(tree.file);
-    }
-    pipeline::typecheck(finalGS, move(resolved), config.opts, workers);
     auto out = finalGS->errorQueue->drainWithQueryResponses();
     finalGS->lspTypecheckCount++;
     finalGS->lspQuery = core::lsp::Query::noQuery();
-    return TypecheckRun{move(out.first), move(affectedFiles), move(finalGS), move(updates), false};
+
+    if (committed) {
+        prodCategoryCounterInc("lsp.updates", "slowpath");
+        return TypecheckRun(move(finalGS), move(out.first), move(affectedFiles), move(updates), false);
+    } else {
+        prodCategoryCounterInc("lsp.updates", "slowpath_canceled");
```

Comment:
I want to track these separately so we understand how much user pain we are saving.